### ISSUE TITLE
JetStream latency spikes with DiscardOld

### ISSUE THREAD

**Issue by delaneyj** [nats-io/nats-server#3948]
## Defect

So I'm hitting a fresh stream with satellite data.  sat.tracking.[id].  Worked great until I set a MaxMsgsPerSubject:100, Discard:nats.DiscardOld .  After the limit I'm seeing very bad spikes that don't recover.  The latencies are in the 10s of seconds even with an embedded nats.  In memory shows the same behavior but with 1/4 the latency (graphs at end).

https://natsio.slack.com/archives/CM3T6T7JQ/p1678300347977169

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included `nats-server -DV` output: 
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats-server` and affected client libraries used:
```bash
[4919] 2023/03/08 11:40:48.220414 [INF] Starting nats-server
[4919] 2023/03/08 11:40:48.220454 [INF]   Version:  2.9.15
[4919] 2023/03/08 11:40:48.220456 [INF]   Git:      [b91fa85]
[4919] 2023/03/08 11:40:48.220457 [DBG]   Go build: go1.19.6
[4919] 2023/03/08 11:40:48.220458 [INF]   Name:     NB2HGZHILGSBJDOWKN4F6JWSHE4H7UR6J5G7NGVGGOFMQSBYO3GSXQF2
[4919] 2023/03/08 11:40:48.220460 [INF]   ID:       NB2HGZHILGSBJDOWKN4F6JWSHE4H7UR6J5G7NGVGGOFMQSBYO3GSXQF2
[4919] 2023/03/08 11:40:48.220472 [DBG] Created system account: "$SYS"
[4919] 2023/03/08 11:40:48.220870 [INF] Listening for client connections on 0.0.0.0:4222
[4919] 2023/03/08 11:40:48.220888 [DBG] Get non local IPs for "0.0.0.0"
[4919] 2023/03/08 11:40:48.220997 [DBG]   ip=172.19.102.157
[4919] 2023/03/08 11:40:48.221023 [INF] Server is ready
[4919] 2023/03/08 11:40:48.221213 [DBG] maxprocs: Leaving GOMAXPROCS=20: CPU quota undefined
```

#### OS/Container environment:
Ubuntu 22.04/WSL

#### Steps or code to reproduce the issue:
```bash
nats s create test --subjects='test.>' --storage file --max-msgs-per-subject=100 --defaults
nats bench --js --pub 1 --multisubject --multisubjectmax 7289 --pubbatch 2000 --msgs 800000 --stream test test.hello
```

#### Actual result:
X=count of messages per subject, Y=latencyInSec

File
![image](https://user-images.githubusercontent.com/438252/223828313-27ad9d32-09e6-42f0-a500-dff802a4b002.png)
Memory
![image](https://user-images.githubusercontent.com/438252/223828374-6ec99d27-4fce-4e80-b6f0-e3e3fb17c1b6.png)

**Comment by codegangsta** [comment]
FYI I tested this on `main`, `2.9.10` and `2.8.6` and experienced similar behavior.

**Comment by jnmoyne** [comment]
@derekcollison this looks very similar to what I have seen, in my case the expiration being because of max age and I exacerbate the problem by having 256 publishers at the same time.

**Comment by delaneyj** [comment]
Confirmed this behavior is fixed in 2.9.16! Thanks @derekcollison
